### PR TITLE
Use storybook's addon-contexts to allow for theme toggling

### PR DESCRIPTION
**Pull request by BPScott**
### WHY are these changes introduced?

Preliminary theming work is rad. Having an easy way to preview it on existing stories would be even radder.

And while I'm playing with contexts, being able to disable react strict mode so I can get accurate counts of how often things get rerendered when debugging would be super useful too.

### WHAT is this pull request doing?

Use storybook's addon-contexts addon to allow for theme toggling, and react strict mode toggling

- Add addon-contexts and remove addon-backgrounds
- Configure context to allow toggling React strict mode on/off
- Configure context to allow configuring theme setups. There is a
choice of default, global theming light mode and global theming dark
mode

#### Themes

Visit a story of a component that is themed, e.g. Banner. Click through the options available in the theme button (the paintbrush in the Storybook topbar, sorry I can't fix the the mystery meat navigation) and marvel at new colors.

Alternatively use this playground that has lots of them on:

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React, {useState, useCallback} from 'react';
import {
  ArrowLeftMinor,
  HomeMajorMonotone,
  OrdersMajorTwotone,
  ProductsMajorTwotone,
  OnlineStoreMajorTwotone,
  CirclePlusOutlineMinor,
} from '@shopify/polaris-icons';
import {
  Page,
  Card,
  List,
  AppProvider,
  Checkbox,
  Frame,
  Banner,
  Stack,
  TextContainer,
  Button,
  Link,
  Badge,
  TopBar,
  ActionList,
  Navigation,
} from '../src';

export function Playground() {
  const [isUserMenuOpen, setIsUserMenuOpen] = useState(false);
  const [isSearchActive, setIsSearchActive] = useState(false);
  const [searchValue, setSearchValue] = useState('');

  const toggleIsUserMenuOpen = useCallback(
    () => setIsUserMenuOpen((isUserMenuOpen) => !isUserMenuOpen),
    [],
  );

  const handleSearchResultsDismiss = useCallback(() => {
    setIsSearchActive(false);
    setSearchValue('');
  }, []);

  const handleSearchChange = useCallback((value) => {
    setSearchValue(value);
    setIsSearchActive(value.length > 0);
  }, []);

  const handleNavigationToggle = useCallback(() => {
    console.log('toggle navigation visibility');
  }, []);

  const userMenuMarkup = (
    <TopBar.UserMenu
      actions={[
        {
          items: [{content: 'Back to Shopify', icon: ArrowLeftMinor}],
        },
        {
          items: [{content: 'Community forums'}],
        },
      ]}
      name="Dharma"
      detail="Jaded Pixel"
      initials="D"
      open={isUserMenuOpen}
      onToggle={toggleIsUserMenuOpen}
    />
  );

  const searchResultsMarkup = (
    <Card>
      <ActionList
        items={[
          {content: 'Shopify help center'},
          {content: 'Community forums'},
        ]}
      />
    </Card>
  );

  const searchFieldMarkup = (
    <TopBar.SearchField
      onChange={handleSearchChange}
      value={searchValue}
      placeholder="Search"
    />
  );

  const topBarMarkup = (
    <TopBar
      showNavigationToggle
      userMenu={userMenuMarkup}
      searchResultsVisible={isSearchActive}
      searchField={searchFieldMarkup}
      searchResults={searchResultsMarkup}
      onSearchResultsDismiss={handleSearchResultsDismiss}
      onNavigationToggle={handleNavigationToggle}
    />
  );

  return (
    <Frame
      topBar={topBarMarkup}
      navigation={
        <Navigation location="/">
          <Navigation.Section
            rollup={{
              after: 3,
              view: 'view',
              hide: 'hide',
              activePath: '/',
            }}
            items={[
              {
                url: '/path/to/place',
                label: 'Home',
                icon: HomeMajorMonotone,
                selected: true,
              },
              {
                url: '/path/to/place',
                label: 'Home 2',
                icon: HomeMajorMonotone,
                secondaryAction: {
                  url: '/admin/orders/add',
                  accessibilityLabel: 'Add an order',
                  icon: CirclePlusOutlineMinor,
                },
              },
              {
                url: '/path/to/place',
                label: 'Orders',
                icon: OrdersMajorTwotone,
                selected: true,
                badge: (
                  <Badge status="new" size="small">
                    46
                  </Badge>
                ),
                subNavigationItems: [
                  {
                    url: '/path/to/place',
                    label: 'All orders',
                    selected: true,
                  },
                  {url: '/path/to/place', label: 'Drafts'},
                  {
                    url: '/path/to/place',
                    label: 'Abandoned checkouts',
                    disabled: true,
                  },
                ],
              },
              {
                url: '/path/to/place',
                label: 'Products',
                icon: ProductsMajorTwotone,
                disabled: true,
              },
            ]}
          />
          <Navigation.Section
            separator
            title="Sales channels"
            items={[
              {
                url: '/path/to/place',
                label: 'Online Store',
                icon: OnlineStoreMajorTwotone,
              },
            ]}
            action={{
              accessibilityLabel: 'Add sales channel',
              icon: CirclePlusOutlineMinor,
              onClick: () => {},
            }}
          />
        </Navigation>
      }
    >
      <Page title="Playground">
        <TextContainer>
          <Banner status="info" title="Info">
            <p>
              Add weights to show accurate rates at checkout and when buying
              shipping labels in Shopify.
            </p>
          </Banner>
          <Banner status="warning" title="Warning">
            <p>
              Add weights to show accurate rates at checkout and when buying
              shipping labels in Shopify.
            </p>
          </Banner>
          <Banner status="critical" title="Critical">
            <p>
              Add weights to show accurate rates at checkout and when buying
              shipping labels in Shopify.
            </p>
          </Banner>
          <Banner status="success" title="Success">
            <p>
              Add weights to show accurate rates at checkout and when buying
              shipping labels in Shopify.
            </p>
          </Banner>
          <Banner title="Neutral">
            <p>
              Add weights to show accurate rates at checkout and when buying
              shipping labels in Shopify.
            </p>
          </Banner>
          <Card
            title="Shipment 1234"
            secondaryFooterActions={[{content: 'Edit shipment'}]}
            primaryFooterAction={{content: 'Add tracking number'}}
          >
            <Card.Section>
              <Banner status="info">
                <p>
                  Add weights to show accurate rates at checkout and when buying
                  shipping labels in Shopify.
                </p>
              </Banner>
              <Banner status="warning">
                <p>
                  Add weights to show accurate rates at checkout and when buying
                  shipping labels in Shopify.
                </p>
              </Banner>
              <Banner status="critical">
                <p>
                  Add weights to show accurate rates at checkout and when buying
                  shipping labels in Shopify.
                </p>
              </Banner>
              <Banner status="success">
                <p>
                  Add weights to show accurate rates at checkout and when buying
                  shipping labels in Shopify.
                </p>
              </Banner>
              <Banner>
                <p>
                  Add weights to show accurate rates at checkout and when buying
                  shipping labels in Shopify.
                </p>
              </Banner>
            </Card.Section>
            <Card.Section title="Items">
              <TextContainer>
                <Stack>
                  <Badge>Fulfilled</Badge>
                  <Badge status="info" progress="partiallyComplete">
                    Published
                  </Badge>
                  <Badge status="warning" progress="partiallyComplete">
                    SSL unavailable
                  </Badge>
                  <Badge status="success" progress="partiallyComplete">
                    Funds recovered
                  </Badge>
                  <Badge progress="incomplete">Unfulfilled</Badge>
                  <Badge progress="partiallyComplete">
                    Partially fulfilled
                  </Badge>
                  <Badge progress="complete">Fulfilled</Badge>
                </Stack>
                <List>
                  <List.Item>1 × Isis Glass, 4-Pack</List.Item>
                  <List.Item>1 × Anubis Cup, 2-Pack</List.Item>
                </List>
              </TextContainer>
            </Card.Section>
          </Card>
        </TextContainer>
      </Page>
    </Frame>
  );
}

```

</details>

#### Strict mode
To test the React strict mode toggle works, use the following sandbox: When in strict mode (the default) clicking the increment button should result in two console logs from the render method (as strict mode [double invokes render() calls in class components](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects).

Click the strict mode context toggle (the speedometer looking thing in the storybook ui, next to the zoom buttons, alas I can't fix the mystery meat navigation). and disable strict mode. Clicking the increment button should now console log one item.

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page, Button} from '../src';

export function Playground() {
  const [count, setCount] = React.useState(1);

  const increment = React.useCallback(() => setCount((cnt) => cnt + 1), []);
  return (
    <Page title="Playground">
      <Button onClick={increment}>Increment</Button>

      <ClassyItem text={count.toString()} />
    </Page>
  );
}

class ClassyItem extends React.Component<any> {
  render() {
    console.log(
      'classy render! (I shall be logged twice in React Strict Mode because it double invokes render() functions',
    );
    return <div>{this.props.text}</div>;
  }
}
```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

### CODE REVIEW

**Comment by BPScott** [comment]
uugh failures.

- TS is failing because the update brought in `@types/webpack-env` and it has a delcaration of NodeModule that disagrees with the one provided by `@shopify/typescript-configs. Compare the Hot interface in https://unpkg.com/browse/@types/webpack-env@1.14.0/index.d.ts vs https://github.com/Shopify/sewing-kit/blob/master/packages/typescript-configs/definitions/module.d.ts. @beefchimi had the same issue with this when he was updating online-store-ui.  @michenly might have some thoughts on how to fix this as she was the mastermind behind the tsconfigs package. Either updating the type to mimic the `webpack-env` package or removing our definition and making sewing-kit depend upon `webpack-env`


- The disabled select story is failing accessibility checks. This is a legit failure that was never picked up before as we unset the body background-color that polaris set, and instead relied upon the color defined in addon-backgrounds shining through from behind the iframe. This meant the a11y tests were unaware of the actual background color. The test passed because it was comparing against a white background, instead of the actual sky light color. Ping @amrocha, @alex-page for thoughts on what to do here to fix that test (that ideally isn't bring back the shitlist for one last hurrah).

**Comment by tmlayton** [comment]
> The disabled select story is failing accessibility checks

Bump up the text darkness imo

**Comment by tmlayton** [comment]
Just poked at it a bit more—the a11y test for disabled select is not exactly right as it calculates the text contrast against the body instead of the select background. Regardless, it still fails with a 2.62 instead of a 2.52. Looks like we’d need to update the color to #647380 for this to pass (4.5) against the body, although it would really be 4.61.

a11y

File: .storybook/config.js [hunk 1]
```diff
@@ -29,27 +30,63 @@ addParameters({
...
+        {name: 'React Strict Mode On', default: true, props: {isStrict: true}},
+        {name: 'React Strict Mode Off', props: {isStrict: false}},
+      ],
+    },
+    {
+      icon: 'paintbrush',
+      title: 'Themes',
+      components: [AppProvider],
+      params: [
+        {name: 'Base Theming', default: true, props: {i18n: enTranslations}},
```
Comment (tmlayton) [inline .storybook/config.js]:
```suggestion
        {name: 'Global Theming Disabled', default: true, props: {i18n: enTranslations}},
```

**Comment by tmlayton** [comment]
> WCAG's own documentation states that disabled UI elements don't need to meet contrast ratios.

I get why it was failing then. The display content is not part of the `select` so a11y tests didn’t know it was supposed to be a disabled element. I added another commit which reverts the text color change and instead sets `aria-disabled` on the content node which passes the a11y tests.

**Comment by BPScott** [comment]
Great sleuthing on the select stuff. To fix the type issue I've reverted back to storybook 5.2.1 (same as master). We're still blocked on updating to newer versions, but we don't need to update right now anyway.

**Comment by BPScott** [comment]
> Also, a special request. It would be incredibly powerful to be able to have a couple of CI checks run with these provider settings in place. In particular, it would be useful to have a non-blocking pally check run for color contrast on both light mode and dark mode. It would also be great to have non-blocking Percy checks for both modes as well. This would help with review, and help us deliver an accessible themed experience. Thoughts on that as a possibility?

Strong agree that being able to test these things would be very powerful. The good news is that contexts works by adding a query string to the preview pane so we can hit https://polaris-react.herokuapp.com/iframe.html?id=playground-playground--playground&contexts=Themes=Global%20Theming%20Enabled%20-%20Light%20Mode to give us the playground in light mode for instance, so these urls are available to be used.

I'm a bit wary of what you mean by "non-blocking check run" do you mean:

- A new check that fails but does not block merging - in which case we train people to ignore failed checks, which I don't think we want to get people in the habit of that.
- Optional steps in the existing check that show stuff in the build logs but don't trigger a failure - in which case that info is easy to overlook and we don't realize there are issues.

Got to admit, I don't really like either of those options, I'd rather stick with failing when we encounter problems and making sure they get fixed.

Adding these checks to the existing pa11y check should be pretty easy - we'd need to rejig [how we built the list of urls to check](https://github.com/Shopify/polaris-react/blob/master/scripts/pa11y.js#L58-L86) to include adding the extra context variations we desire.

Adding these checks to Percy will be a bit more tricky as it currently doesn't support those arbitrary extra params for each story. I'll raise an issue with them - supporting an official addon that is a general-case solution to their existing "rtl" solution seems like something they'd go for.

**Comment by danrosenthal** [comment]
I'm in agreement with you on the `non-blocking` stuff. Let's just make the checks fail.

**Comment by BPScott** [comment]
Percy issue created: https://github.com/percy/percy-storybook/issues/146

**Pull request merged**